review

emsiish opened the pull request.

emsiish committed:
basic questionnaire collecting answers

emsiish committed:
Update .gitignore

emsiish committed:
removed .idea folder

emsiish committed:
navbar, progress and products layout

emsiish committed:
separated project into backend and frontend; added full controller to entity functionality to doctor

emsiish committed:
basic backend

emsiish committed:
removed .idea

emsiish committed:
connecting backend and frontend

emsiish committed:
sending tags to backend and updating user tags

Teddy15 commented on backend/.mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.9.5/apache-maven-3.9.5-bin.zip
```
Този файл не ти е нужен в хранилището, слагай такива файлове в git ignore.

Teddy15 commented on backend/mvnw:
```diff
@@ -0,0 +1,308 @@
+#!/bin/sh
```
Този файл не ти е нужен в хранилището, слагай такива файлове в git ignore.

Teddy15 commented on backend/mvnw.cmd:
```diff
@@ -0,0 +1,205 @@
+@REM ----------------------------------------------------------------------------
```
Този файл не ти е нужен в хранилището, слагай такива файлове в git ignore.

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/DoctorController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @PostMapping
+    public DoctorResource addDoctor(@RequestBody DoctorResource doctor) {
+        return doctorService.addDoctor(doctor);
```
Един съвет - добави си логове, ще улеснят процеса на дебъгване. Логовете обаче нека имат някакви уместни и смислени съобщения а не просто "test1", "method addDoctor" или нещо подобно
Виж slf4j dependency-то и как се използва.

Teddy15 replied:
Важи за всеки контролер + сървис клас

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/resources/DoctorRatingResource.java:
```diff
@@ -0,0 +1,11 @@
+package com.example.skincarerecs.controller.resources;
```
Както говорихме, преименувай си на DoctorRatingDto класа. Като това важи за всички resource класове, които имаш. Пакетът също го прекръсти на dto.
Друго важно нещо - помисли кои полета не трябва да са празни (ако има такива).
Разгледай @NotNull, @NotBlank, @NotEmpty анотациите. (javax.validation.constraints им е пакетът, виж кое dependency трябва да се добави, ако има нужда също) 
Също така, ако искаш някакви custom валидации да имаш за имейл/телефон/каквото и да е - възможно е.
Искам добре да помислиш тези твои Data transfer objects от какви задължителни полета  се нуждаят : )
Пиши ми, ако имаш нужда от помощ

Teddy15 replied:
Важи за всичките класове. Няма да копирам коментара.
Също, за да се валидират тези анотации накрая, в съответния контролер клас преди някой @RequestBody параметър трябва да добавиш @Valid анотацията (javax.validation)

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/entity/DoctorRating.java:
```diff
@@ -0,0 +1,27 @@
+    private Long id;
+
+    private String comment;
+    private int rating;
```
Тоест, мога да дам само 1 или 2 звезди, например, не мога да дам 1,5 като рейтинг на някого?

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/entity/ProductRating.java:
```diff
@@ -0,0 +1,27 @@
+    private Long id;
+
+    private String comment;
+    private int rating;
```
Тоест, мога да дам само 1 или 2 звезди, например, не мога да дам 1,5 като рейтинг на някого?

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/entity/User.java:
```diff
@@ -0,0 +1,35 @@
+    private Long id;
+
+    private String name;
+    private String username;
```
Имейл няма ли да пазиш? Това е обвързано с аутентикацията ти всъщност. Трябва да прецениш имейл ли е по-добре или username. Кое какви функции има, как те лимитира, ако те лимитира и т.н.

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/mapper/DoctorMapper.java:
```diff
@@ -0,0 +1,19 @@
+
+@Mapper
+public interface DoctorMapper {
+    DoctorMapper DOCTOR_MAPPER = Mappers.getMapper(DoctorMapper.class);
```
Защо правиш извикване към това factory?

Teddy15 replied:
Същия въпрос се отнася към всички mapper-и

Teddy15 replied:
Оставям го тук, за да го видиш също: относно въпросът с мапърите, може да изредиш кои останали мапъри искаш да ползваш в DoctorMapper например.
Това става като напишеш @Mapper(uses = {TestMapper.class})
Другото, което е, виж какво е componentModal = "spring". Ето така се ползва: @Mapper(componentModel="spring").

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/repository/TagRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+@Repository
+public interface TagRepository extends JpaRepository<Tag, Long> {
+    List<Tag> findAllByNameIn(List<String> tagsNames);
```
Защо си кръстила този метод findAllByNameIn, а не просто findAllByName ?

Teddy15 replied:
Ненужен въпрос, на второ четене разбрах защо. findAllByNameIn не е достатъчно описателно според мен, ако за пръв път четеш кода може да те обърка
findAllByNameInList или нещо такова е по-добре според мен

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/DoctorRatingService.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface DoctorRatingService {
+    DoctorRatingResource addDoctorRating(Long doctorId, DoctorRatingResource doctorRating);
+    List<DoctorRatingResource> getAllDoctorRatings(Long doctorId);
```
Не мисля, че някъде ще ти трябва да листваш всички дадени оценки на докторите от всички потребители? 
Поне не се сещам, ако бъркам, моля поправи ме

emsiish replied:
Ами мислих, че ще е хубаво примерно като цъкнеш някой доктор и да се display-ват оценките и коментарите дадени за него

Teddy15 replied:
Ясно, разбрах, ок

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/DoctorService.java:
```diff
@@ -0,0 +1,17 @@
+
+    DoctorResource getDoctorById(Long id);
+
+    DoctorResource updateDoctor(Long id, DoctorResource doctor);
```
Помисли дали същото DTO трябва да ползваш при update. Тоест, помисли кои данни трябва да се ъпдейтват и кои не

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/ProductRatingService.java:
```diff
@@ -0,0 +1,15 @@
+    ProductRatingResource getProductRatingById(Long productId, Long id);
+
+    //?? do i need it?
+    ProductRatingResource updateProductRating(Long productId, Long id, ProductRatingResource productRating);
```
Кой и кога редактира оценките на продуктите?

emsiish committed:
updated gitignore, added validation, logs and fixed mappers and data types, removed unused update functions

emsiish committed:
fixed user and product adding, filtering products based on tags, displaying doctors, showing and adding ratings to products

Teddy15 commented on backend/pom.xml:
```diff
@@ -0,0 +1,107 @@
+			<artifactId>flyway-core</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.flywaydb</groupId>
```
Добавила си dependency-то, но не виждам миграции за базата ти
Ще е хубаво да добавиш няколко миграции за създаването на таблциите
Ако имаш нужда от помощ, пиши ми на лично, за да ги видим

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/CorsConfig.java:
```diff
@@ -0,0 +1,17 @@
+public class CorsConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
```
Само вметка, че евентуално, когато настроиш аутентикацията и ауторизацията на приложнеието ти, този конфиг и логиката за него най-вероятно ще трябва да се премести. Когато си готова да го направиш, ако имаш затруднения, пиши ми, но исках да си наясно

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/resources/UserResource.java:
```diff
@@ -0,0 +1,15 @@
+    private String username;
+    private String password;
+
+    private List<String> tags_names;
```
Придържай се имената на методите и променливите ти да следват camelCase конвенцията
Тоест, tags_names -> tagNames

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/resources/ProductResource.java:
```diff
@@ -0,0 +1,19 @@
+    private String image;
+    private String price;
+
+    private List<String> tags_names;
```
Придържай се имената на методите и променливите ти да следват camelCase конвенцията
Тоест, tags_names -> tagNames

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/resources/ProductRatingResource.java:
```diff
@@ -0,0 +1,12 @@
+    private Long id;
+    private String comment;
+    private int rating;
+    private String product_name;
```
Придържай се имената на методите и променливите ти да следват camelCase конвенцията
Тоест, product_name -> productName

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/controller/resources/DoctorResource.java:
```diff
@@ -0,0 +1,12 @@
+public class DoctorResource {
+    private Long id;
+    private String name;
+    private String telephone;
```
Може би по-добре phoneNumber да се казва полето

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/entity/Doctor.java:
```diff
@@ -0,0 +1,26 @@
+    private Long id;
+
+    private String name;
+    private String telephone;
```
тук също phoneNumber

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/TagService.java:
```diff
@@ -0,0 +1,12 @@
+    TagResource getTagById(Long id);
+
+    //do i need it??
+    TagResource updateTag(Long id, TagResource tag);
```
Кой и кога ъпдейтва таговете? Ако за момента таговете ще можеш само ръчно да ги вкарваш, колко точно от изписаната функционалност ти трябва?
Евентуално аз си представям, че някой с роля Админ би трябвало да редактира такава информация и да има такива права. Остави си функционалността, но не мисля, че ти е нужна в момента при положение, че не ти е описано нищо за Админ панел в Заданието на ДР. Мисълта ми е да не губиш време с излишни за момента добавки.

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/UserService.java:
```diff
@@ -0,0 +1,14 @@
+    UserResource updateUser(Long id, UserResource user);
+    void deleteUser(Long id);
+
+    List<UserResource> getAllUsers();
```
На кого му е нужно да вижда списък с всички потребители? Кой и кога би направил такава заявка?

Teddy15 commented on backend/src/main/java/com/example/skincarerecs/service/impl/DoctorRatingServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        User userEntity = userRepository.findByUsername(doctorRating.getUsername()).orElseThrow();
+        Doctor doctorEntity = doctorRepository.findById(doctorId).orElseThrow();
+
+        doctorRatingEntity.setUser(userEntity);
```
Прегледай ми коментарите за Mapping анотацията и като цяло за mapper-ите ти.
В момента, в който направиш нужните промени такива добавки в сървис класовете ти вече няма да са ти нужни и логиката ти ще бъде една идея по-подредена.

Teddy15 replied:
Важи за всички подобни методи в Сървис класовете ти

Teddy15 commented on backend/src/main/resources/application.properties:
```diff
@@ -0,0 +1,11 @@
+spring.jpa.hibernate.ddl-auto=update
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true
```
какво прави това пропърти?

Teddy15 commented on backend/src/main/resources/application.properties:
```diff
@@ -0,0 +1,11 @@
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+server.port=8080
```
това си е дефолтния порт по прицнип, не е нужно да го сетваш
Реакт  приложението ти върви на 3000, така че няма да имаш проблеми

Teddy15 commented on backend/src/test/java/com/example/skincarerecs/BackendApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
```
В някакъв момент, ще е хубаво да добавиш поне unit tests към приложението ти.
Важи и за Бекенда и за Фронтенда.

Teddy15 commented on frontend/package-lock.json:
Добави го git ignore-a на проекта този файл.

Teddy15 commented on frontend/public/logo512.png:
В някакъв момент изтрий тези дефолтни изображения

Teddy15 commented on frontend/tailwind.config.js:
```diff
@@ -0,0 +1,9 @@
+/** @type {import('tailwindcss').Config} */
+module.exports = {
+  content: ["./src/**/*.{js,jsx,ts,tsx}",],
```
Планираш да пишеш и JavaScript, и TypeScript?

Teddy15 commented on frontend/src/components/styles.css:
```diff
@@ -0,0 +1,110 @@
+/* styles.css */
```
Ако вече си си избрала някакви дефолтни цветове, шрифтове и тн, добави си ги в App.css, защото те са общи за целия app

Teddy15 commented on frontend/src/components/Welcome.js:
```diff
@@ -0,0 +1,19 @@
+// src/components/WelcomePage.js
```
Може би Welcome не е най-идеалното име за този компонент

Teddy15 commented on frontend/src/components/Navbar.js:
```diff
@@ -0,0 +1,18 @@
+// src/components/Navbar.js
```
Хубаво  е в папката ти components/ да имаш отделни папки за компонентите.
Тоест, Navbar компонетът ти ще се намира в components/navbar
Защо? Защото понякога е нужно компонентите да си имат специфичен css само за тях. В такъв случай, ето така биха ти изглеждали пътищата:
components/navbar/Navbar.js
components/navbar/navbar.css
Изцяло от гледна точка на подреденост и четимост е коментарът

Teddy15 commented on frontend/src/components/Products.js:
```diff
@@ -0,0 +1,47 @@
+
+    useEffect(() => {
+        // Fetch products from the backend when the component mounts
+        axios.get('http://localhost:8080/api/v1/products')
```
Такива ендпоийнти е добре да са изведени в отделен файл и да са константи

Teddy15 commented on frontend/src/components/Products.js:
```diff
@@ -0,0 +1,47 @@
+                                <img
+                                    src={product.image}
+                                    alt={`Product ${product.id}`}
+                                    className="mb-2 w-full h-auto" // Adjust max height as needed
```
Такива коментари са ненужни

Teddy15 replied:
Важи  за всичките ти компоненти
За да не копирам коментара на много места го пиша

Teddy15 commented on frontend/src/components/Questionnaire.js:
```diff
@@ -0,0 +1,81 @@
+// src/components/QuestionPage.js
```
Компонентът ти се казва Questionaire, но експортваш QuestionPage?

Teddy15 commented on frontend/src/components/Results.js:
```diff
@@ -0,0 +1,22 @@
+// src/components/ResultsPage.js
```
Компонентът ти се казва Results, но експортваш ResultsPage?

Teddy15 commented on frontend/src/components/Navbar.js:
```diff
@@ -0,0 +1,18 @@
+// src/components/Navbar.js
+import React from 'react';
```
Искам да прочетеш за js и jsx файловете как се ползват, по какво се различават и да видиш дали не е нужно да добавиш jsx файлове

emsiish committed:
fixed styling and file extensions

emsiish committed:
displaying ratings of the doctors

emsiish committed:
removed redundant files/methods, fixed validation, renaming, endpoints as constants

emsiish committed:
basic auth

emsiish committed:
finished auth, getting current user from jwt

emsiish committed:
sending tags to backend and updating user tags

updated gitignore, added validation, logs and fixed mappers and data types, removed unused update functions

fixed user and product adding, filtering products based on tags, displaying doctors, showing and adding ratings to products

fixed styling and file extensions

displaying ratings of the doctors

removed redundant files/methods, fixed validation, renaming, endpoints as constants

basic auth

finished auth, getting current user from jwt

emsiish committed:
store userID in token, fix pop-up div, register/login links

emsiish committed:
fixed routing, question styling and made questions mandatory

emsiish committed:
questions fetched from backend

emsiish committed:
db migrations, bug fixes, removing redundant code

emsiish committed:
added jwt expiration alert and image to doctors

emsiish committed:
updated doctor migrations

emsiish committed:
improved readability and removed redundant files

emsiish committed:
Merge branch 'main' of https://github.com/emsiish/Skin-care-recs

emsiish committed:
removed redundant files

emsiish committed:
bug fix

emsiish committed:
Update README.md

emsiish committed:
[ADD] tests for doctor service

emsiish committed:
[FIX] jwt expiration time and format of avg rating

emsiish committed:
[ADD] tests for doctor rating and user service

emsiish committed:
[ADD] tests for controllers and product service

emsiish committed:
[ADD] error handling

emsiish committed:
[FIX] made api url env variable

emsiish committed:
[ADD] azure blob storage for doctor images

emsiish committed:
[ADD] icons next to average rating for doctor

emsiish committed:
[FIX] styling

emsiish committed:
[ADD] exception handler for io exception

emsiish committed:
[FIX] images urls in migrations

emsiish committed:
[ADD] dockerfile and actions

emsiish committed:
[UPDATE] Build and Run Actions with maven

emsiish committed:
[FIX] dynamic env.properties for testing

emsiish committed:
[FIX] import of env.properties changed to optional

emsiish committed:
[FIX] Specified PathVariables for github actions

emsiish committed:
[DELETE] redundant test class

emsiish committed:
[DELETE] redundant steps

emsiish committed:
[ADD] dockerfile and actions for frontend

emsiish committed:
[FIX] starting action

emsiish committed:
[ADD] logout button

emsiish committed:
Merge branch 'review'

emsiish committed:
[UPDATE] README.md
